### PR TITLE
respect projection ordering in results

### PR DESCRIPTION
**Pull request by sbaldwin-rs**
Ensures when explicit projects are used, the returned result set is ordered as the projections are.

### CODE REVIEW

File: src/main/java/com/rockset/jdbc/RocksetResultSet.java [hunk 1]
```diff
@@ -1579,22 +1595,23 @@ private static List<Column> getColumns(QueryResponse response) throws SQLExcepti
...
               continue;
             }
-            JsonNode value = field.getValue();
+            JsonNode value = root.get(fieldName);
             Column.ColumnTypes type = Column.ColumnTypes.fromValue(value.getNodeType().toString());
-            // Skip over the fields with null type unless all values for that
-            // field are null
-            if (type.equals(Column.ColumnTypes.NULL) && i != response.getResults().size() - 1) {
-              continue;
-            }
```
Comment (notnoop) [inline src/main/java/com/rockset/jdbc/RocksetResultSet.java:1597]: We may need something more subtle here. If the first row has null but second has a string, the column type should be string. I think this code will infer null now.

File: src/main/java/com/rockset/jdbc/RocksetResultSet.java [hunk 2]
```diff
@@ -1567,55 +1573,49 @@ private static void log(String msg) {
...
-            fieldNames.add(fieldName);
-          }
         }
-      } else if (response.getColumnFields() != null && response.getColumnFields().size() > 0) {
+
+        // If we know the desired field ordering through explicit projection, use that ordering.
+        // Otherwise, just iterate over in arbitrary ordering
+        Collection<String> fields = orderedFields.isEmpty() ? fieldTypes.keySet() : orderedFields;
+        for(String field : fields){
+          out.add(new Column(field, fieldTypes.get(field)));
```
Comment (haneeshr) [inline src/main/java/com/rockset/jdbc/RocksetResultSet.java:1615]: If a field is null in all rows, then it wouldn't be in `fieldTypes` as they get skipped on line 1599 (previously this was handled on like 1595). Idk if `new Column(field, null)` will interpret it as NULL type - if not this needs to be handled

Comment (sbaldwin-rs) [reply src/main/java/com/rockset/jdbc/RocksetResultSet.java:1615]: This is handled by the column conversion, I added a test case to make sure

Comment (haneeshr) [reply src/main/java/com/rockset/jdbc/RocksetResultSet.java:1615]: what do you mean by "column conversion"?

Comment (sbaldwin-rs) [reply src/main/java/com/rockset/jdbc/RocksetResultSet.java:1615]: https://github.com/rockset/rockset-java-client/blob/dac440d557c24025b6b69add829feaca82029b96/src/main/java/com/rockset/jdbc/RocksetResultSet.java#L1641

**Review by haneeshr — APPROVED** [review]

File: src/main/java/com/rockset/jdbc/RocksetResultSet.java (hunk 2 above)
Comment (haneeshr) [reply src/main/java/com/rockset/jdbc/RocksetResultSet.java:1615]: Thanks! I figured the column conversion part but I incorrectly understood that `fieldTypes` wouldn't have an entry for the field until the last row based on the previous implementation which is now being handled differently by tracking the NULL type fields as we see and overriding them when we encounter a non-null type.

**Pull request merged**